### PR TITLE
Added unit tests for unlink() on HasMany association

### PR DESCRIPTION
**Pull request by half2me**
Added unit tests for unlink() on HasMany

### CODE REVIEW

**Comment by half2me** [comment]
fixes #9324

File: tests/TestCase/ORM/Association/HasManyTest.php [hunk 1]
```diff
@@ -643,4 +643,75 @@ protected function assertSelectClause($expected, $query)
...
+        $new = $this->article->get(2, ['contain' => 'Articles']);
+        $this->assertCount(0, $new->articles, 'DB should be clean');
+        $this->assertSame(4, $this->author->find()->count(), 'Authors should still exist');
+        $this->assertSame(3, $this->article->find()->count(), 'Articles should still exist');
+    }
+
+    /**
+     * Tests that unlink will work without contain() being called
+     */
+    public function testUnlinkWithoutContain()
```
Comment (markstory) [inline tests/TestCase/ORM/Association/HasManyTest.php]: I'll be removing this test as it is invalid. The method signature for unlink() requires that `$targetEntities` be an array, which a Query object is not.

File: tests/TestCase/ORM/Association/HasManyTest.php [hunk 2]
```diff
@@ -643,4 +643,75 @@ protected function assertSelectClause($expected, $query)
...
+        $assoc = $this->author->hasMany('Articles', [
+            'sourceTable' => $this->author,
+            'targetTable' => $this->article
+        ]);
+
+        $entity = $this->author->get(1, ['contain' => 'Articles']);
+        $initial = $entity->articles;
+        $this->assertCount(2, $initial);
+
+        $this->author->unlink($entity, $entity->tags);
```
Comment (markstory) [inline tests/TestCase/ORM/Association/HasManyTest.php]: I'm assuming this was an unintentional mistake, that I'll fix.

**Comment by markstory** [comment]
Thanks for the tests :smile:

**Comment by half2me** [comment]
@markstory these unit tests don't show an error I was having. The unit test for calling unlink() with an empty array actually produces a bad SQL query. Please take a look at my issue which was closed. #9324

**Comment by markstory** [comment]
@half2me I did see bad queries being generated. The `[]` case was generating empty `WHERE` clauses which are not valid. These errors are no longer emitted as the method returns early now. 

The test case passing a `Query` object in could not stay in the test suite as it generates a warning which aborts the method call preventing a query from ever being run.

**Pull request merged**